### PR TITLE
Enabled IRF and KWT routing restart option

### PR DESCRIPTION
**Pull request by nmizukami**
1. Added read/write statements for channel flow states for both IRF (UH routed future flow for each segment) and KWT (wave characteristics for each segment) routing.    

2. Changed use of pointers to allocatable array for KWT state variables (KROUTE%KWAVE)

3. control file change. added restart file name.

In addition,
1.  code style changes (indentation, cases), some comments edits, and removal of unnecessary statements.

### CODE REVIEW

**Comment by nmizukami** [comment]
Thanks for reviewing, @andywood 

> ctrl-file:
is it worth considering if we should standardize runoff units to mm/hr? it's more intuitive when scanning data. similarly for precip and et, basically all moisture fluxes.

See line 220-231 in runoff_route.f90. 
The code checks length (depth) unit and time unit for runoff and converts depth unit to meter and time unit to sec. Reason for the conversion is the final unit of streamflow is cubic meter per sec. There is no option for the other units for streamflow.   For now, meter or millimeter is allowed for depth and sec, day,  or year are allowed. Any combinations of those depth-time units are ok (e.g., m/day, m/hr, m/yr, mm/day, ...).   Otherwise you get error.  

This way the code provide some flexibility for units (but not extensive, no English units).  If only one unit is allowed, user might need extra process (unit conversion). For example,  If you have VIC daily output in mm/day in old archive, you would need to covert it to mm/s or rerun VIC with mm/hr.

**Comment by nmizukami** [comment]
> also, state file can be both input and output - should it have its own category here?

The next pull request is for separate input/output state file specifications in ctl file.  For this PR, it writes and reads with the same file name (so always overwrites state file after run, which is not good).

**Comment by nmizukami** [comment]
Please ignore `scripts/Calc_hru_wgtavg_nc.py`.   The latest change in this file has been already merged by pull request #14

File: route/build/Makefile [hunk 1]
```diff
@@ -82,7 +82,8 @@ DATATYPES = \
 UTILS = \
 		nr.f90 \
 		nrutil.f90 \
-		ascii_util.f90
+		ascii_util.f90\
+    public_var.f90
```
Comment (jhamman) [inline route/build/Makefile]: nit: can you make this indentation match the above lines?

File: route/build/src/irf_route.f90 [hunk 2]
```diff
@@ -346,7 +318,7 @@ subroutine get_upsbas_qr(nSeg,iEns, &     ! input
 !    print *,'Working on iSeg, ID, # of ups. basins= ', iSeg, NETOPO(iSeg)%REACHID, nUpsBas
     if (nUpsBas > 0) then ! if segment is not located in headwater basin
       do iUpsbas=1,nUpsBas
-        jUpsBas = NETOPO(iSeg)%UREACHI(iUpsBas)    ! index of the upstream upstream basin
+        jUpsBas = NETOPO(iSeg)%UREACHI(iUpsBas)    ! index of the imediate upstream upstream reach
```
Comment (jhamman) [inline route/build/src/irf_route.f90]: imediate->immediate

File: route/build/src/kwt_route.f90 [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
 integer(i4b),parameter  :: ixPrint = -9999  ! the desired reach (set to negative to avoid any printing)
 real(dp),parameter      :: verySmall=tiny(1.0_dp)  ! a very small number
-contains
 
+contains
 
  ! *********************************************************************
- ! new subroutine: define processing order for the individual
+ ! subroutine: define processing order for the individual
  !                 stream segments in the river network
```
Comment (jhamman) [inline route/build/src/kwt_route.f90]: fix indentation in comment

File: route/build/src/kwt_route.f90 [hunk 4]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
 
+   ! initialize error control
+   ierr=0; message='QROUTE_RCH/'
+   ! ----------------------------------------------------------------------------------------
+   ! (0) INITIALIZE POINTERS
+   ! ----------------------------------------------------------------------------------------
+   if(INIT) then
+     INIT=.false.
+     !NULLIFY(Q_JRCH,TENTRY,T_EXIT,FROUTE,NEW_WAVE)
+     !deallocate(Q_JRCH,TENTRY,T_EXIT,FROUTE,NEW_WAVE)
```
Comment (jhamman) [inline route/build/src/kwt_route.f90]: remove commented out code

File: route/build/src/irf_route.f90 [hunk 5]
```diff
@@ -346,7 +318,7 @@ subroutine get_upsbas_qr(nSeg,iEns, &     ! input
 !    print *,'Working on iSeg, ID, # of ups. basins= ', iSeg, NETOPO(iSeg)%REACHID, nUpsBas
```
Comment (jhamman) [inline route/build/src/irf_route.f90]: remove commented out code

File: route/build/src/kwt_route.f90 [hunk 6]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
+      ierr=20; message=trim(message)//'KROUTE is not associated'; return
+    else
+   ! if(allocated(KROUTE(IENS,JRCH)%KWAVE)) then
+      deallocate(KROUTE(IENS,JRCH)%KWAVE, STAT=ierr)
+   !   if (ierr.ne.0) then
+   !     print *, '% CANNOT DEALLOCATE SPACE (FOR SOME UNKNOWN REASON)... TRY AND NULLIFY!'
+   !     !NULLIFY(KROUTE(IENS,JRCH)%KWAVE)
+   !     deallocate(KROUTE(IENS,JRCH)%KWAVE)
+   !   endif
+   ! endif
```
Comment (jhamman) [inline route/build/src/kwt_route.f90]: remove commented out code

File: route/build/src/kwt_route.f90 [hunk 7]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
+    ! (when JRCH becomes IR routed points will be stripped out & the structures updated again)
+    KROUTE(IENS,JRCH)%KWAVE(0:NR)%QF=Q_JRCH(0:NR); KROUTE(IENS,JRCH)%KWAVE(NR+2:NQ2+1)%QF=Q_JRCH(NR+1:NQ2)
+    KROUTE(IENS,JRCH)%KWAVE(0:NR)%TI=TENTRY(0:NR); KROUTE(IENS,JRCH)%KWAVE(NR+2:NQ2+1)%TI=TENTRY(NR+1:NQ2)
+    KROUTE(IENS,JRCH)%KWAVE(0:NR)%TR=T_EXIT(0:NR); KROUTE(IENS,JRCH)%KWAVE(NR+2:NQ2+1)%TR=T_EXIT(NR+1:NQ2)
+    KROUTE(IENS,JRCH)%KWAVE(0:NR)%RF=FROUTE(0:NR); KROUTE(IENS,JRCH)%KWAVE(NR+2:NQ2+1)%RF=FROUTE(NR+1:NQ2)
+    KROUTE(IENS,JRCH)%KWAVE(0:NQ2+1)%QM=-9999
+    ! implement water use
+    !IF (NUSER.GT.0.AND.UCFFLAG.GE.1) THEN
+      !CALL EXTRACT_FROM_RCH(IENS,JRCH,NR,Q_JRCH,T_EXIT,T_END,TNEW)
+    !ENDIF
```
Comment (jhamman) [inline route/build/src/kwt_route.f90]: remove commented out code

File: route/build/src/kwt_route.f90 [hunk 8]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
+   INTEGER(I4B)                                :: NN            ! # non-routed particles
+   REAL(DP),DIMENSION(2)                       :: TNEW          ! start/end of time step
+   REAL(DP),DIMENSION(1)                       :: QNEW          ! interpolated flow
+   ! (4) housekeeping
+   REAL(DP)                                    :: Q_END         ! flow at the end of the timestep
+   REAL(DP)                                    :: TIMEI         ! entry time at the end of the timestep
+   TYPE(FPOINT),allocatable,DIMENSION(:),SAVE  :: NEW_WAVE      ! temporary wave
+   LOGICAL(LGT),SAVE                           :: INIT=.TRUE.   ! used to initialize pointers
+   ! random stuff
+   CHARACTER(LEN=256)                          :: CMESSAGE      ! error message for downwind routine
```
Comment (jhamman) [inline route/build/src/kwt_route.f90]: @nmizukami - can you add some inline comments on github describing where changes were made in the `QROUTE_RCH` subroutine. The indentation changes are making this tough to review.

File: route/build/src/kwt_route.f90 [hunk 9]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
+      endif
+      allocate(KROUTE(IENS,JRCH)%KWAVE(0:NN),STAT=IERR)  ! again, the zero element for the last routed point
+      if(ierr/=0)then; message=trim(message)//'problem allocating space for KROUTE'; return; endif
+      ! copy data back to the wave structure and deallocate space for the temporary wave
+      KROUTE(IENS,JRCH)%KWAVE(0:NN) = NEW_WAVE(0:NN)
+      DEALLOCATE(NEW_WAVE,STAT=IERR)
+      if(ierr/=0)then; message=trim(message)//'problem deallocating space for NEW_WAVE'; return; endif
+    endif  ! (if JRCH is the last reach)
+    ! get size of wave number for a reach
+    return
```
Comment (jhamman) [inline route/build/src/kwt_route.f90]: this seems odd to me. What is happening here?

File: route/build/src/kwt_route.f90 [hunk 10]
```diff
@@ -861,7 +853,8 @@ SUBROUTINE QEXMUL_RCH(IENS,JRCH,T0,T1,&            ! input
  ! ----------------------------------------------------------------------------------------
  IF(INIT) THEN
   INIT=.FALSE.
-  NULLIFY(USFLOW,NEW_WAVE,QD_TEMP,TD_TEMP)
+  !NULLIFY(USFLOW,NEW_WAVE,QD_TEMP,TD_TEMP)
+  !deallocate(USFLOW,NEW_WAVE,QD_TEMP,TD_TEMP)
```
Comment (jhamman) [inline route/build/src/kwt_route.f90]: remove commented out code

File: route/build/src/route_runoff.f90 [hunk 11]
```diff
@@ -115,13 +86,12 @@ program route_runoff
 integer(i4b)               :: iRch                ! index in reach structures
 !integer(i4b)               :: jRch               ! index in reach structures
 ! define simulated runoff data at the HRUs
-real(dp)                   :: dt                  ! time step (seconds)
 real(dp)                   :: dTime               ! time variable (in units units_time)
+real(dp)                   :: TB(2)
```
Comment (jhamman) [inline route/build/src/route_runoff.f90]: add a comment for this variable.

File: route/build/src/route_runoff.f90 [hunk 12]
```diff
@@ -503,15 +433,18 @@ program route_runoff
...
+  RPARAM(:)%R_WIDTH = wscale * sqrt(RPARAM(:)%TOTAREA)  ! channel width (m)
+  RPARAM(:)%R_MAN_N = mann_n                            ! Manning's "n" paramater (unitless)
+
   ! define processing order of the reaches
   call reachorder(nSegRoute, ierr, cmessage); call handle_err(ierr, cmessage)
   !! check
   !do iRch=1,nSegRoute
   ! jRch = NETOPO(iRch)%RHORDER
   ! write(*,'(a,1x,2(i4,1x),f20.2)') 'iRch, NETOPO(jRch)%DREACHI, RPARAM(jRch)%TOTAREA/1000000._dp  = ', iRch, NETOPO(jRch)%DREACHI, RPARAM(jRch)%TOTAREA/1000000._dp
   !end do
```
Comment (jhamman) [inline route/build/src/route_runoff.f90]: remove commented out code

File: route/build/src/write_simoutput.f90 [hunk 13]
```diff
@@ -133,19 +144,153 @@ end subroutine defvar
...
+   call defvar('TI',         'entry time of a wave in reach',             'sec', (/sSeg_DimName,sWav_DimName,sEns_dimName/),    nf90_double, ierr, cmessage)
+   call defvar('TR',         'time when a wave is expected to exit reach','sec', (/sSeg_DimName,sWav_DimName,sEns_dimName/),    nf90_double, ierr, cmessage)
+   call defvar('RF',         'routing flat (1 if wave has exited reach)', '-',   (/sSeg_DimName,sWav_DimName,sEns_dimName/),    nf90_int,    ierr, cmessage) 
+ endif
+ ! end definitions
+ ierr = nf90_enddef(ncid)
+ if(ierr/=0)then; message=trim(message)//trim(nf90_strerror(ierr)); return; endif
+ ! close NetCDF file
+ ierr = nf90_close(ncid)
+ if(ierr/=0)then; message=trim(message)//trim(nf90_strerror(ierr)); return; endif
```
Comment (jhamman) [inline route/build/src/write_simoutput.f90]: can you also add a few basic global attributes to the state file?

See http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#description-of-file-contents

File: route/build/src/kwt_route.f90 [hunk 14]
```diff
@@ -241,67 +242,67 @@ SUBROUTINE REACH_LIST(NRCH,NTOTAL,ierr,message)
...
 
  ! *********************************************************************
- ! new subroutine: route kinematic waves through the river network
+ ! subroutine: route kinematic waves through the river network
  ! *********************************************************************
- SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
+ subroutine QROUTE_RCH(IENS,JRCH,    & ! input: array indices
                        ixOutlet,     & ! input: index of the outlet reach
                        T0,T1,        & ! input: start and end of the time step
+                       MAXQPAR,      & ! input: maximum number of particle in a reach
```
Comment (nmizukami) [inline route/build/src/kwt_route.f90]: Added this for output argument

File: route/build/src/kwt_route.f90 [hunk 15]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
- INTEGER(I4B)                                :: NN      ! # non-routed particles
- REAL(DP),DIMENSION(2)                       :: TNEW    ! start/end of time step
- REAL(DP),DIMENSION(1)                       :: QNEW    ! interpolated flow
- ! (4) housekeeping
- REAL(DP)                                    :: Q_END   ! flow at the end of the timestep
- REAL(DP)                                    :: TIMEI   ! entry time at the end of the timestep
- TYPE(FPOINT), DIMENSION(:), POINTER, SAVE   :: NEW_WAVE    ! temporary wave
- LOGICAL(LGT),SAVE                           :: INIT=.TRUE. ! used to initialize pointers
- ! random stuff
- INTEGER(I4B),PARAMETER                      :: MAXQPAR=20  ! maximum number of particles
```
Comment (nmizukami) [inline route/build/src/kwt_route.f90]: This MAXQPAR is now declared as output intent

File: route/build/src/kwt_route.f90 (hunk 9 above)
Comment (nmizukami) [reply route/build/src/kwt_route.f90]: I think the comment can be removed (caused confusion?)

File: route/build/src/kwt_route.f90 [hunk 16]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
- ! initialize error control
- ierr=0; message='QROUTE_RCH/'
- !print*, 'jRch, ixOutlet = ', jRch, ixOutlet
- ! ----------------------------------------------------------------------------------------
- ! (0) INITIALIZE POINTERS
- ! ----------------------------------------------------------------------------------------
- IF(INIT) THEN
-  INIT=.FALSE.
-  NULLIFY(Q_JRCH,TENTRY,T_EXIT,FROUTE,NEW_WAVE)
- ENDIF
```
Comment (nmizukami) [inline route/build/src/kwt_route.f90]: I changed pointers (T_EXIT, Q_JRCH, TENTRY, FROUTE, NEW_WAVE) to allocatable array. See declarations above.   So I could have removed INITIALIZE POINTERS sections.

File: route/build/src/kwt_route.f90 (hunk 8 above)
Comment (nmizukami) [reply route/build/src/kwt_route.f90]: Sorry, I changed indentation so whole thing was changed....  I also changed case (from upper to lower) for some.   only a few real changes. I put inline comments.   1. change pointers to allocatable array, and 2. move MAXQPAR to input argument.

File: route/build/src/kwt_route.f90 [hunk 17]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
+        deallocate(KROUTE(IENS,JRCH)%KWAVE,STAT=IERR)
+        if(ierr/=0)then; message=trim(message)//'problem deallocating space for KROUTE'; return; endif
+      endif
+      allocate(KROUTE(IENS,JRCH)%KWAVE(0:0),STAT=ierr)     
+      if(ierr/=0)then; message=trim(message)//'problem allocating space for KROUTE(IENS,JRCH)%KWAVE(1)'; return; endif
+      KROUTE(IENS,JRCH)%KWAVE(0)%QF=-9999 
+      KROUTE(IENS,JRCH)%KWAVE(0)%TI=-9999 
+      KROUTE(IENS,JRCH)%KWAVE(0)%TR=-9999 
+      KROUTE(IENS,JRCH)%KWAVE(0)%RF=.False. 
+      KROUTE(IENS,JRCH)%KWAVE(0)%QM=-9999
```
Comment (nmizukami) [inline route/build/src/kwt_route.f90]: Line 483 - 493 are added

File: route/build/src/kwt_route.f90 [hunk 18]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
- IF (.NOT.ASSOCIATED(KROUTE(IENS,JRCH)%KWAVE))THEN
-  ierr=20; message=trim(message)//'KROUTE is not associated'; return
- ENDIF
- IF(ASSOCIATED(KROUTE(IENS,JRCH)%KWAVE)) THEN
-  DEALLOCATE(KROUTE(IENS,JRCH)%KWAVE, STAT=IERR)
-  IF (IERR.NE.0) THEN
-   print *, '% CANNOT DEALLOCATE SPACE (FOR SOME UNKNOWN REASON)... TRY AND NULLIFY!'
-   NULLIFY(KROUTE(IENS,JRCH)%KWAVE)
-  ENDIF
- ENDIF
```
Comment (nmizukami) [inline route/build/src/kwt_route.f90]: Line 545- 552 was changed  (see Line 553-566 in new)

File: route/build/src/kwt_route.f90 (hunk 6 above)
Comment (nmizukami) [reply route/build/src/kwt_route.f90]: Note Line 553-566 are updated.   I did not change logic.   I can clean up.

File: route/build/src/kwt_route.f90 (hunk 4 above)
Comment (nmizukami) [reply route/build/src/kwt_route.f90]: I think I can remove this section 0.

File: route/build/src/kwt_route.f90 [hunk 19]
```diff
@@ -408,216 +409,218 @@ SUBROUTINE QROUTE_RCH(IENS,JRCH,    & ! input: array indices
...
+   USE reachparam
+   USE reachstate
+   USE reach_flux
+   implicit none
+   ! Input
+   INTEGER(I4B), INTENT(IN)                    :: IENS          ! ensemble member
+   integer(i4b), intent(in)                    :: ixOutlet      ! index of the outlet reach
+   INTEGER(I4B), INTENT(IN)                    :: JRCH          ! reach to process
+   REAL(DP), INTENT(IN)                        :: T0,T1         ! start and end of the time step (seconds)
+   INTEGER(I4B), INTENT(IN)                    :: MAXQPAR       ! maximum number of particles
```
Comment (nmizukami) [inline route/build/src/kwt_route.f90]: I moved MAXQPAR here, but I define this variable in public_var.f90. so maybe I could remove.

File: route/build/src/kwt_route.f90 (hunk 10 above)
Comment (nmizukami) [reply route/build/src/kwt_route.f90]: This section was removed since these variables are deallocated in the end of this subroutine

**Comment by msaharia** [comment]
@nmizukami Should I merge this with the develop branch? I am running into merge conflicts with `scripts/Calc_hru_wgtavg_nc.py`

File: route/build/src/write_simoutput.f90 (hunk 13 above)
Comment (nmizukami) [reply route/build/src/write_simoutput.f90]: Added global attributes for model outputs and restarts (contents could be modified better)

**Pull request closed without merging**